### PR TITLE
fix: skip finalizing anonymous class fields in FinalizeLocalVariables

### LINKED ISSUE #181: FinalizeLocalVariables fails when processing fields of anonymous classes
Greetings. Thanks everyone for your work in this useful library.

We've found a issue with the [FinalizeLocalVariables](https://docs.openrewrite.org/recipes/staticanalysis/finalizelocalvariables) recipe, by which fields of anonymous classes that are reassigned in overriden methods are finalized, generating compilation problems in processed files.

## What version of OpenRewrite are you using?
staticAnalysis 1.0.7
RecipeBom 2.3.0
RewriteMavenPlugin 5.7.1

## How are you running OpenRewrite?
Through Maven plugin, in a multi-module project.

## What is the smallest, simplest way to reproduce the problem?
The following test exposes the problem.
```java
package Test;

import javafx.beans.property.IntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;

    class Test {

      private final IntegerProperty cellCount = new SimpleIntegerProperty(this, "cellCount", 0) {
        private int oldCount = 0;

        @Override
        protected void invalidated() {
          final int currentCellCount = this.get();

          final boolean countChanged = this.oldCount != currentCellCount;
          if (countChanged) {
            this.oldCount = currentCellCount;
          }
        }
      };
    }
}
```

## What did you expect to see?
No changes should be applied to the previous code.

## What did you see instead?
**oldCount** var was finalized, regardless of assignment made in _invalidated()_.

```java
package Test;

import javafx.beans.property.IntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;

    class Test {

      private final IntegerProperty cellCount = new SimpleIntegerProperty(this, "cellCount", 0) {
        private final int oldCount = 0;

        @Override
        protected void invalidated() {
          final int currentCellCount = this.get();

          final boolean countChanged = this.oldCount != currentCellCount;
          if (countChanged) {
            this.oldCount = currentCellCount;
          }
        }
      };
    }
}
```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
Yes, although the fix we have is to identify fields of inner classes and simply skip them. This is a workaround to avoid breaking compilation in processed projects.

workaround

### PR DESCRIPTION
**Pull request by AlejandroBertolo**
## What's changed?
Variables identified as fields of anonymous classes will be skipped in **FinalizeLocalVariables** recipe.

## What's your motivation?
Fixes #181 

## Anything in particular you'd like reviewers to focus on?
There is perhaps a better way of identifying anonymous classes.

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [ ] I've added the license header to any new files through `./gradlew licenseFormat`
 - no new files were added
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by AlejandroBertolo** [comment]
Found a simpler way to identify anonymous class parents, as seen in MissingOverrideAnnotation recipe.
Should do the trick.

File: src/test/java/org/openrewrite/staticanalysis/FinalizeLocalVariablesTest.java [hunk 1]
```diff
@@ -394,48 +396,78 @@ class Person {
...
+
+    @Test
+    @Issue("https://github.com/openrewrite/rewrite-static-analysis/issues/181")
+    void shouldNotFinalizeVariablesWhichAreAssignedInAnonymousClasses() {
+        this.rewriteRun(
+          // language=java
+          java(
+            """
+              import javafx.beans.property.IntegerProperty;
+              import javafx.beans.property.SimpleIntegerProperty;
```
[the remark this answers is not in the document]
Comment (AlejandroBertolo) [reply src/test/java/org/openrewrite/staticanalysis/FinalizeLocalVariablesTest.java]: done :) 
Btw, I can't seem to apply proper format using the default/project profiles of intellij, I get a lot of non-related diffs. Sorry about that.

File: src/test/java/org/openrewrite/staticanalysis/FinalizeLocalVariablesTest.java [hunk 2]
```diff
@@ -394,48 +396,70 @@ class Person {
...
-                    static {
-                        switch (variable) {
-                          case 0:
-                              int notFinalized = 0;
-                          default:
-                              notFinalized = 1;
-                        }
+          //language=java
+          java("""
+            class A {
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/staticanalysis/FinalizeLocalVariablesTest.java:428]:
```suggestion
          java(
            """
            class A {
```

**Review by timtebeek — APPROVED** [review]

**Comment by timtebeek** [comment]
Thanks for the quick work on this!

File: src/test/java/org/openrewrite/staticanalysis/FinalizeLocalVariablesTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (timtebeek) [reply src/test/java/org/openrewrite/staticanalysis/FinalizeLocalVariablesTest.java]: No worries; we're looking to automate that part of the review anyway; for now I think I've fixed it. In short we place braces and parentheses each on a new line, as well as the opening/closing of text blocks. Then we indent with two spaces for every level in the tests. That way you know what to aim for, even if you IDE disagrees.

**Pull request merged**